feat: export transformWithEsbuild

bluwy opened the pull request:
<!-- Thank you for contributing! -->

### Description

Expose `transformWithEsbuild` from the `vite` package. Also fixes `tsconfigRaw` merging for `transformWithEsbuild` and perf improvements.

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

- `transformWithEsbuild` is currently already exposed under [server](https://vitejs.dev/guide/api-javascript.html#vitedevserver), but I don't see it coupled to the server. It could be used as a standalone.
- This also allows `vite-plugin-svelte` to use it for preprocessing a Svelte file script tag with `lang="ts"`
- Do we want to deprecate `server.transformWithEsbuild`?

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

bluwy committed:
feat: export transformWithEsbuild

bluwy committed:
chore: add tests

dominikg commented on packages/playground/tsconfig-json/__tests__/tsconfig-json.spec.ts:
```diff
@@ -14,3 +18,38 @@ test('should respected each `tsconfig.json`s compilerOptions', () => {
   expect(browserLogs).toContain('data setter in NestedWithExtendsBase')
 })
+
+describe.only('transformWithEsbuild', () => {
```
.only ?

dominikg replied:
```suggestion
describe('transformWithEsbuild', () => {
```

bluwy committed:
fix: test typo

Co-authored-by: Dominik G. <dominik.goepel@gmx.de>

benmccann commented:
Should the "b" be capitalized? I.e. `transformWithEsBuild`?

bluwy commented:
I think `transformWithEsbuild` reads fine too as `esbuild` is usually all lower caps, but in code we capitalise the first `e`.

yyx990803 reviewed (dismissed):
Yes I think we can deprecate `server.transformWithEsbuild` with this.

bluwy committed:
chore: deprecate server.transformWithEsbuild

bluwy commented:
Added deprecation jsdoc for `server.transformWithEsbuild` and removed docs for it.

Currently, the vue plugin is still using `server.transformWithEsbuild`. With the new API, it seems that it could benefit from a sizable refactor so I'm not touching it for now. It should still work as before.

https://github.com/vitejs/vite/blob/e623b4a34878d8f2a88f515ea798d814790742ee/packages/plugin-vue/src/main.ts#L261

patak-js commented on docs/guide/api-javascript.md:
```diff
@@ -87,16 +87,6 @@ interface ViteDevServer {
    * Apply Vite built-in HTML transforms and any plugin HTML transforms.
    */
   transformIndexHtml(url: string, html: string): Promise<string>
-  /**
-   * Util for transforming a file with esbuild.
-   * Can be useful for certain plugins.
-   */
-  transformWithEsbuild(
-    code: string,
-    filename: string,
-    options?: EsbuildTransformOptions,
-    inMap?: object
-  ): Promise<ESBuildTransformResult>
```
I think we should add the description of the new free `transformWithEsbuild` function to the docs

bluwy replied:
I can add it, but there's a lot of other APIs not documented currently in the [Javascript API docs](https://vitejs.dev/guide/api-javascript.html) too. So I thought `transformWithEsbuild` should follow suit.

patak-js replied:
Since plugin-svelte would use this one and it was documented before, I think it is better in this case to keep a minimal reference to it in the docs

bluwy replied:
@patak-js I've added the docs. Along the way I also found that we need to export the `EsbuildTransformOptions` type, so I did that as well, otherwise that type would be `TransformOptions_2`

bluwy committed:
docs: add transformWithEsbuild
